File: LeetCode/Easy/0219-contains-duplicate-ii/0219-contains-duplicate-ii.py
from collections import deque


class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        end = len(nums)
        windowq = deque(nums[:k+1])
        count_dict = Counter(windowq)
        if len(set(windowq)) != len(windowq):
            return True

        for i in range(k+1, end):
            byebye = windowq.popleft()
            count_dict[byebye] -= 1
            if count_dict[nums[i]] == 1:
                return True
            count_dict[nums[i]] += 1
            windowq.append(nums[i])

        return False

File: LeetCode/Easy/0219-contains-duplicate-ii/0219-contains-duplicate-ii.kt
class Solution {
    fun containsNearbyDuplicate(nums: IntArray, k: Int): Boolean {
        val window = mutableSetOf<Int>()
        
        nums.forEachIndexed { i, num ->
            if (num in window) return true
            window.add(num)
            if (window.size > k) {
                window.remove(nums[i - k])
            }
        }
        
        return false
    }
}